kube-ctrl-mgr: enable secure port 10257

sttts opened the pull request:
This PR enables authn+authz (delegated to the kube-apiserver) and the secure port 10257 for the kube-controller-manager. In addition, the insecure port is disabled.

Moreover, it adds integration test coverage for the `--port` and `--secure-port` flags, plus the testserver infrastructure to tests flags in general inside integration tests.

```release-note
Enable secure serving on port 10257 to kube-controller-manager (configurable via `--secure-port`). Delegated authentication and authorization are to be configured using the same flags as for aggregated API servers. Without configuration, the secure port will only allow access to `/healthz`.

```

sttts committed:
kube-controller-manager: add test server and test serving

andrewsykim commented on pkg/master/ports/ports.go:
```diff
@@ -41,4 +41,10 @@ const (
 	// ProxyHealthzPort is the default port for the proxy healthz server.
 	// May be overridden by a flag at startup.
 	ProxyHealthzPort = 10256
+	// InsecureKubeControllerManagerPort is the default port for the controller manager status server.
```
s/InsecureKubeControllerManagerPort/KubeControllerManagerPort/

sttts replied:
fixed

andrewsykim commented on pkg/master/ports/ports.go:
```diff
@@ -41,4 +41,10 @@ const (
+	// InsecureKubeControllerManagerPort is the default port for the controller manager status server.
+	// May be overridden by a flag at startup.
+	KubeControllerManagerPort = 10257
+	// InsecureCloudControllerManagerPort is the default port for the cloud controller manager server.
```
s/InsecureCloudControllerManagerPort/CloudControllerManagerPort

sttts replied:
fixed

sttts committed:
kube-controller-manager: enable delegated authz/authn if secure port is enabled

sttts committed:
kube-controller-manager: enable secure ports 10257, deprecate insecure port

sttts committed:
kube-controller-manager: enable secure loopback

sttts commented:
Split out the cloud-ctrl-mgr changes into https://github.com/kubernetes/kubernetes/pull/67069

sttts commented:
@liggitt @deads2k please take a look especially at the second last commit with the tests whether this is the behaviour we want: we enable the secure port by default, exclude `/healthz` from authn/authz, and forbid every other path. Hence, we start securely by default, and the way forward is clear by setting up delegated authn/authz correctly.

sttts commented:
/retest

nikhita commented:
/remove-area custom-resources

yue9944882 commented:
/assign

sttts commented:
@yue9944882 rebased. ptal.

sttts commented:
/retest

sttts commented:
/retest

deads2k commented on pkg/master/ports/ports.go:
```diff
@@ -41,4 +42,7 @@ const (
 	ProxyHealthzPort = 10256
+	// KubeControllerManagerPort is the default port for the controller manager status server.
+	// May be overridden by a flag at startup.
+	KubeControllerManagerPort = 10257
```
We're claiming a port.  @liggitt does that require an api review or something?

liggitt replied:
I don't actually know

deads2k commented:
Wiring pull based that uses blocks we've already merged.

/lgtm
/hold

holding pending the question about port reservation for @liggitt

sttts commented:
/retest

liggitt commented:
re: https://github.com/kubernetes/kubernetes/pull/64149/commits/c12d77d0da9bce6b42d139ef24ce7343f3783fa6#r213671335

I don't really know how the other ports were determined. This keeps up the "increment by one" pattern, for better or worse. Nothing showed up in the IANA registered port list for this port.

ccing @kubernetes/api-reviewers in case there are other considerations/processes we should be aware of

otherwise this seems reasonable to me. will hold through end of the day tomorrow in case someone else wants to chime in.

sttts commented:
/retest

sttts committed:
kube-controller-manager: disable authn/z on insecure port

This is the old behaviour and we did not intent to change it due to enabled authn/z in general.
As the kube-apiserver this sets the "system:unsecured" user info.

liggitt commented:
the `kube-controller-manager: disable authz on insecure port` commit preserves current behavior for the unsecured healthz port

sttts commented:
Rebased, advanced auditing GA merged.

sttts commented:
/retest

sttts commented:
@liggitt lgty and ready to unhold?

liggitt commented:
/lgtm
/hold cancel
